src/libtorch.cc

```diff
@@ -0,0 +1,1089 @@
+      "failed reporting batch request statistics");
+}
+
+void
+ModelInstanceState::Execute(
+    std::vector<TRITONBACKEND_Response*>* responses,
+    const uint32_t response_count,
+    std::vector<torch::jit::IValue>* input_tensors,
+    std::vector<torch::Tensor>* output_tensors)
+{
+  torch::jit::IValue model_outputs_;
+
+  try {
+    model_outputs_ = torch_model_->forward(*input_tensors);
+    if (model_outputs_.isTuple()) {
+      auto model_outputs_tuple = model_outputs_.toTuple();
+      for (auto& m_op : model_outputs_tuple->elements()) {
+        output_tensors->push_back(m_op.toTensor());
+      }
+    } else {
+      auto model_output_tensor = model_outputs_.toTensor();
+      output_tensors->push_back(model_output_tensor);
+    }
+  }
+  catch (std::exception& ex) {
+    SendErrorForResponses(
+        responses, response_count,
+        TRITONSERVER_ErrorNew(
+            TRITONSERVER_ERROR_INTERNAL,
+            ("PyTorch execute failure: " + std::string(ex.what())).c_str()));
+  }
+}
+
+void
+ModelInstanceState::SetInputTensors(
+    size_t total_batch_size, TRITONBACKEND_Request** requests,
+    const uint32_t request_count,
+    std::vector<TRITONBACKEND_Response*>* responses,
+    BackendInputCollector* collector, std::vector<const char*>* input_names,
+    std::vector<torch::jit::IValue>* input_tensors, bool* cuda_copy)
+{
+  const int max_batch_size = model_state_->MaxBatchSize();
+
+  // All requests must have equally-sized input tensors so use any
+  // request as the representative for the input tensors.
+  uint32_t input_count;
+  RESPOND_ALL_AND_RETURN_IF_ERROR(
+      responses, request_count,
+      TRITONBACKEND_RequestInputCount(requests[0], &input_count));
+  input_tensors->resize(input_count);
+  for (uint32_t input_idx = 0; input_idx < input_count; input_idx++) {
+    TRITONBACKEND_Input* input;
+    RESPOND_ALL_AND_RETURN_IF_ERROR(
+        responses, request_count,
+        TRITONBACKEND_RequestInputByIndex(requests[0], input_idx, &input));
+
+    const char* input_name;
+    TRITONSERVER_DataType input_datatype;
+    const int64_t* input_shape;
+    uint32_t input_dims_count;
+    RESPOND_ALL_AND_RETURN_IF_ERROR(
+        responses, request_count,
+        TRITONBACKEND_InputProperties(
+            input, &input_name, &input_datatype, &input_shape,
+            &input_dims_count, nullptr, nullptr));
+
+    input_names->emplace_back(input_name);
+
+    // The shape for the entire input patch, [total_batch_size, ...]
+    std::vector<int64_t> batchn_shape(
+        input_shape, input_shape + input_dims_count);
+    if (max_batch_size != 0) {
+      batchn_shape[0] = total_batch_size;
+    }
+
+    // The input must be in contiguous CPU/GPU memory.
+    const int64_t batchn_byte_size = GetByteSize(input_datatype, batchn_shape);
+
+    std::initializer_list<BackendMemory::AllocationType> alloc_perference;
+    if (device_.is_cpu()) {
+      alloc_perference = {BackendMemory::AllocationType::CPU_PINNED_POOL,
```

Comment:
If the model is expecting input in CPU memory then don't use pinned. That is only if we needed to first stage some CPU data before transferring it to GPU.

Reply:
Fixed